### PR TITLE
feat(render): lazy frame-start + per-row column-range diff (#172)

### LINKED ISSUE #172: Lazy frame-start + per-row column-range diff (port from OpenTUI)
## What to build

Two related ANSI-byte optimizations from `docs/research/OPENTUI_COMPARISON.md` §A1 + §A3, ported from OpenTUI's renderer. Combined, they save 50–90% of bytes per streaming tick on typical chat updates and eliminate idle-tick noise.

## Source of truth

- `docs/research/OPENTUI_COMPARISON.md` §A1, §A3, §D
- OpenTUI reference: `zig/renderer.zig:1247-1393`, `zig/renderer.zig:1304-1307`, `zig/renderer.zig:1331-1349`
- SumoTUI today: `src/sumo-tui/render/diff.ts` (114 lines), `src/sumo-tui/runtime/terminal-controller.ts:182-190`

### Part 1 — Lazy frame-start (no-op suppression)

**Today:** `writeFramePatches` always wraps patches in `\x1b[?2026h … \x1b[?2026l` plus a cursor-show, even when `patches.length === 0`. ~12+ ANSI bytes plus cursor reposition per idle tick.

**Fix:** in `runtime/terminal-controller.ts writeFramePatches`, early-return when `patches.length === 0 && !cursorMoved`. Skip cursor restore when its position is unchanged from `lastEmittedCursor` (new cache field).

### Part 2 — Per-row column-range diff

**Today:** `render/diff.ts:39-46` regenerates the **whole row** ANSI even when one cell changed. Wastes bytes on cursor blinks and single-cell streaming updates.

**Fix:**
1. Widen `FrameDiffPatch` to `{ row, startCol, ansi, type }` by extending row-comparison to return leftmost/rightmost differing columns
2. In `terminal-controller.writeFramePatches`, when `startCol > 0` skip `\x1b[K` and use `\x1b[r;c+1H<ansi>`
3. Add `lastEmittedCursor` cache field on `TerminalController`

## Acceptance criteria

- [ ] No-op tick (no diff, no cursor move) emits **zero bytes** to stdout
- [ ] One-cell change in row N emits a patch with `startCol > 0` and uses `\x1b[N;C+1H` not `\x1b[N;1H` + clear-line
- [ ] Cursor reposition is emitted only when `(row, col)` differs from `lastEmittedCursor`
- [ ] Existing diff tests still pass; new tests cover:
  - One-cell-change asserting `startCol > 0`
  - No-op tick asserting zero writes
  - Idle frame loop (10 ticks, no state change) asserts zero stdout writes
- [ ] `pnpm test`, `pnpm test:integration`, `pnpm exec tsc --noEmit`, `pnpm build` all green
- [ ] Manual smoke: streaming chat session shows no perceptible flicker; idle session shows ~0 wakes per second in `SUMO_TUI_DEBUG=1` overlay

## Effort

~1.5 days including tests.

## Files touched

- `src/sumo-tui/render/diff.ts` (widen `FrameDiffPatch`, add column-range logic)
- `src/sumo-tui/render/diff.test.ts` (extend coverage)
- `src/sumo-tui/runtime/terminal-controller.ts` (early-return, cursor cache, partial-row patches)
- `src/sumo-tui/runtime/terminal-controller.test.ts` (no-op tick test, cursor cache test)

## Trade-offs

- Diff and writer become slightly more complex; tests cover the new shape
- Regression risk: easy to introduce subtle off-by-one in column-range; tests must include edge cases (start of row, end of row, full-row change still works)
- Not a correctness fix — purely a perf/byte optimization. Skip if perf isn't user-perceptible

## Framing

This is **candidate work, not a v1 blocker.** Per `docs/SUMO_TUI_AUDIT_V2.md` §5 STOP-list, items become work only when dogfood data justifies them. File this if you observe streaming flicker, high stdout volume in profiling, or terminal-side scrollback noise. Otherwise sit on it.

## Related

- Research: `docs/research/OPENTUI_COMPARISON.md` §A1 + §A3 + §D (highest-impact single port)
- PR #171 — research doc
- Audit v2: `docs/SUMO_TUI_AUDIT_V2.md` (#166)
- Doesn't conflict with #168 (scroll smoothness, tactical coalescing) or #165 (selection)

### PR DESCRIPTION
**Pull request by dhruvkelawala**
## Summary

Ports two related ANSI-byte optimizations from OpenTUI's renderer (`anomalyco/opentui` `zig/renderer.zig` 1247-1393, 1304-1307, 1331-1349). See `docs/research/OPENTUI_COMPARISON.md` §A1 + §A3 + §D.

Closes #172.

## Perf benchmark

Real measurements via `node scripts/bench-diff-bytes.mjs` (committed in this PR for reproducibility, runs from any checkout). Compares OLD path (always full-row repaint + always emit `\x1b[?2026h` wrapper + always re-emit cursor) against the NEW path (column-range diff + lazy frame-start + `lastEmittedCursor` cache) on the same input frames.

| Scenario | Old | New | Reduction |
|---|---:|---:|---:|
| Cursor blink, single cell in 200-col row | **240B** | **39B** | **−83.8%** |
| Streaming append 1 char at end of 200-col row | **238B** | **37B** | **−84.5%** |
| Streaming append 10 chars | **238B** | **45B** | **−81.1%** |
| Idle tick (no change, cursor unchanged) | **29B** | **0B** | **−100%** |
| 10 idle ticks (same cursor) | **290B** | **29B** | **−90.0%** |
| Cursor moves 1 col (no patches) | 29B | 29B | 0% (genuine move, no save possible) |
| Full-row change in 40×200 frame | 239B | 239B | 0% (cursor cache doesn't apply when patches exist) |
| Streaming chunk: 5 mid-row chars + cursor move | **240B** | **44B** | **−81.7%** |

The wins concentrate exactly where they matter:
- **Cursor blink** (terminal repaints once a second by default): −84%.
- **Streaming chunks** (every assistant token): −81% to −85%.
- **Idle frames** (most ticks during a long-running session): −100% per tick after the first.

The "cursor moves 1 col" row stays flat at 29B because the cursor genuinely moved and nothing else changed — that's the lower bound for an actual cursor reposition. Full-row changes show 0% reduction because patches are emitted (cursor cache doesn't apply, by design — see "Cursor reposition correctness" below).

### Diff layer (`src/sumo-tui/render/diff.ts`)

- `FrameDiffPatch` widened with `startCol: number`. Full-row patches (and scroll patches) keep `startCol: 0`; partial patches set the column where the change begins.
- New `rowChangeRange()` finds leftmost + rightmost differing cells. Backs up over wide-char continuation cells so the slice always begins at the head of a glyph.
- `changedRowPatches()` emits a column-range patch when the change starts past column 0; otherwise falls back to full-row repaint (legacy behaviour, safe with `\x1b[K`).
- `cellRowToAnsi()` now thin-wraps `cellRowSliceToAnsi()`.

### Writer (`src/sumo-tui/render/ansi-writer.ts`)

- New exported `cellRowSliceToAnsi(buffer, row, startCol, endCol)` emits SGR-coalesced ANSI for a column slice.

### Terminal controller (`src/sumo-tui/runtime/terminal-controller.ts`)

- `writeFramePatches()` lazy frame-start: when there are no patches AND the cursor would land on its last-emitted position, emit zero bytes (no `\x1b[?2026h` wrapper, no cursor reposition).
- `lastEmittedCursor` cache field tracks the last cursor position so redundant repositions are suppressed when there are also no patches.
- Partial patches (`startCol > 0`) skip the trailing `\x1b[K` so cells right of the change region survive untouched.
- `exitTerminal()` resets `lastEmittedCursor = null` so re-entering altscreen always re-emits cursor.
- `TerminalPatch.startCol` is optional (defaults to 0) so existing callers don't need to change.

### Cursor reposition correctness

Cursor-write elision is **only safe when `patches.length === 0`**. When patches are emitted, every `\x1b[r;c+1H<ansi>` in the loop physically moves the terminal cursor. Skipping the final cursor reposition would leave the visible caret parked at the end of the last patch instead of at the intended cursor position. The implementation always re-emits the cursor sequence when patches were written. (Surfaced as a P1 review comment from `chatgpt-codex-connector`; addressed in commit `66b6023`.)

## Tests

- `src/sumo-tui/render/diff.test.ts` — added 3 new column-range cases (right-side change, single mid-row cell, change-starts-at-col-0 fallback) plus updated existing tests for the new `startCol` field.
- `src/sumo-tui/runtime/terminal-controller.test.ts` — added 6 new cases:
  - partial-row patch skips `\x1b[K`
  - no-op tick emits 0 bytes
  - cursor-unchanged emits 0 bytes when patches empty
  - cursor reposition always re-emitted when patches exist (correctness gate)
  - cursor-only emit for patches=0 + cursor moved
  - fresh session re-emits cursor after exitTerminal

## Verification

- `pnpm test`: 552/552 ✅
- `pnpm test:integration`: 18/18 ✅
- `pnpm exec tsc --noEmit`: clean ✅

## Conflict notes

Touches `terminal-controller.ts` (new `writeFramePatches` body, new field). PR #165 also touches this file but in a different region (`MOUSE_SGR_ENABLE_SEQUENCE` constant), so a merge is mechanical. PR #170 reads `patches.length` from the new `writeFramePatches` contract via `sumo-interactive-mode.ts` `logDiagnostic("render_patches", { patchCount })` — that contract is preserved (patches array is still the input).

## Independence from other open PRs

Can land independently of #165, #167, #168, #169, #170, #175. None of them touch the diff layer or the `writeFramePatches` body. Order doesn't matter — light merge resolution if any.

Closes #172.

### CODE REVIEW

**Comment by dhruvkelawala** [comment]
@codex Review

File: src/sumo-tui/runtime/terminal-controller.ts [hunk 1]
```diff
@@ -180,12 +196,32 @@ export class TerminalSessionOwner {
...
+			output += `\x1b[${patch.row + 1};${startCol + 1}H${patch.ansi}`;
+			// Only full-row patches benefit from `\x1b[K` (clear-to-end-of-line):
+			// they overwrite every cell on the row anyway. Partial-row patches
+			// (`startCol > 0`) MUST NOT emit `\x1b[K` or they would wipe correct
+			// cells to the right of the change region.
+			if (startCol === 0) output += "\x1b[K";
+		}
+		if (cursor && cursorMoved) {
+			output += `\x1b[${cursor.row + 1};${cursor.col + 1}H\x1b[?25h`;
+			this.lastEmittedCursor = { row: cursor.row, col: cursor.col };
```
[the remark this answers is not in the document]
Comment (dhruvkelawala) [reply src/sumo-tui/runtime/terminal-controller.ts]: Confirmed bug — fixed in commit 66b6023. Cursor-write elision is now restricted to patches.length === 0; when any patches were emitted, the cursor reposition is always re-emitted. Added a regression test (re-emits cursor after every patch frame even when the logical cursor is unchanged). Bench numbers updated in PR body — slightly narrower savings (~−84% vs −89%) for streaming-append cases; full-row change drops from −5% back to 0%. The idle-tick and same-position-no-patches cases still hit 0B.

File: scripts/bench-diff-bytes.mjs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ * Run: `node scripts/bench-diff-bytes.mjs`
+ *
+ * No-op tick measurements use the actual `TerminalSessionOwner` lazy
+ * frame-start. Per-frame measurements use the diff layer directly to keep
+ * the comparison apples-to-apples between paths.
+ */
+import { createJiti } from "@mariozechner/jiti";
+
+const jiti = createJiti(import.meta.url, { moduleCache: false, tryNative: false });
+const root = "/Users/dhruvkelawala/development/sumocode/src/sumo-tui";
```
[the remark this answers is not in the document]
Comment (dhruvkelawala) [reply scripts/bench-diff-bytes.mjs]: Fixed in commit 66b6023. Now resolves paths via fileURLToPath(import.meta.url) + resolve(SCRIPT_DIR, '..', 'src', 'sumo-tui'), so the script runs from any checkout location.

**Comment by dhruvkelawala** [comment]
@codex review

File: src/sumo-tui/runtime/terminal-controller.ts [hunk 3]
```diff
@@ -180,12 +196,37 @@ export class TerminalSessionOwner {
...
+			if (startCol === 0) output += "\x1b[K";
+		}
+		// Cursor-write elision is safe ONLY for true no-op frames. When
+		// `patches.length > 0`, every `\x1b[r;c+1H<ansi>` in the loop above
+		// physically moved the terminal cursor; skipping the reposition here
+		// would leave the visible caret parked at the end of the last patch
+		// instead of at `cursor`. Always re-emit when patches were written.
+		if (cursor && (patches.length > 0 || cursorMoved)) {
+			output += `\x1b[${cursor.row + 1};${cursor.col + 1}H\x1b[?25h`;
+			this.lastEmittedCursor = { row: cursor.row, col: cursor.col };
```
[the remark this answers is not in the document]
Comment (dhruvkelawala) [reply src/sumo-tui/runtime/terminal-controller.ts:228]: Confirmed bug — fixed in commit 63d22cd. When patches are emitted with a null cursor, lastEmittedCursor is now explicitly set to null so the next frame with a logical cursor matching the previously-cached position re-emits its reposition correctly. Added a regression test ("invalidates the cursor cache when emitting patches with a null cursor") that drives the exact A → B → C sequence: seed cache, fire patches with null cursor, then assert that a frame matching the original cursor still emits the cursor sequence.
